### PR TITLE
Release resources

### PR DESCRIPTION
**Pull request by jessitron**
The objective is for the cached directory manager to attach a release step to the context, and have it invoked automatically when the invocation finishes.

### CODE REVIEW

File: src/internal/transport/AbstractRequestProcessor.ts [hunk 1]
```diff
@@ -170,7 +179,8 @@ export abstract class AbstractRequestProcessor implements RequestProcessor {
                 .catch(err => {
                     this.handleCommandError(err, command, ci, ctx, callback);
                 });
-        } catch (err) {
+        } catch
+            (err) {
```
Comment (ddgenome) [inline src/internal/transport/AbstractRequestProcessor.ts]: Why the newline?

File: src/internal/metadata/metadataReading.ts [hunk 2]
```diff
@@ -22,7 +22,7 @@ export function metadataFromInstance(h: any): CommandHandlerMetadata | EventHand
     if (isEventHandlerMetadata(h)) {
         md = addName(h);
     } else if (isCommandHandlerMetadata(h)) {
-        md = addName(h);
+        md = addName(h); // JESS: this seems mighty incomplete to me.
```
Comment (ddgenome) [inline src/internal/metadata/metadataReading.ts]: A note for later?

File: src/internal/invoker/resourceRecovery.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+                    .then(() => f2.how()
+                        .catch(error => {
+                            logger.warn("Failed to release resource %s: %s", f2.what, error);
+                        }))
+                , what: f1.what + " and " + f1.what,
+            };
+        }
+
+        const releaseEverything = ctx.releaseSteps
+            .reduce(both, { how: () => Promise.resolve(), what: "inconceivable" });
```
Comment (ddgenome) [inline src/internal/invoker/resourceRecovery.ts]: Does a simple `map` not work nicely with promises?

File: src/server/BuildableAutomationServer.ts [hunk 4]
```diff
@@ -214,13 +214,8 @@ export class BuildableAutomationServer extends AbstractAutomationServer {
...
 
     private enrichContext(ctx: HandlerContext): HandlerContext {
-        return {
-            teamId: ctx.teamId,
-            correlationId: ctx.correlationId,
-            invocationId: ctx.invocationId,
-            messageClient: ctx.messageClient,
-            graphClient: ctx.graphClient ? ctx.graphClient : this.graphClient,
-        };
+        ctx.graphClient = ctx.graphClient || this.graphClient;
```
Comment (ddgenome) [inline src/server/BuildableAutomationServer.ts]: Before, you were creating a new object, now you are returning the same, possibly modified object. Is this distinction important here?

File: src/internal/invoker/resourceRecovery.ts (hunk 3 above)
Comment (ddgenome) [reply src/internal/invoker/resourceRecovery.ts]: Looking at the tests, it appears this allows for continuing should something fail. Is that correct?

Comment (jessitron) [reply src/internal/invoker/resourceRecovery.ts]: yes, that is the goal. and it puts them in sequence... although I'm not certain that's necessary, it is kinda reassuring

File: src/server/BuildableAutomationServer.ts (hunk 4 above)
Comment (jessitron) [reply src/server/BuildableAutomationServer.ts]: extremely. the context is now mutated and carries information back to the automation.
hmm, which brings up a point. I could put the mutated data in the invoking function, and instead pass a function inside the context which releases it. That would be safer. Then the context would close over some mutable data, but not hold it. Is that better?

File: src/internal/metadata/metadataReading.ts (hunk 2 above)
Comment (jessitron) [reply src/internal/metadata/metadataReading.ts]: yeah, I suspect that whole case doesn't actually work. which is not related to this PR.

File: src/server/BuildableAutomationServer.ts (hunk 4 above)
Comment (ddgenome) [reply src/server/BuildableAutomationServer.ts]: I don't know.

**Comment by ddgenome** [comment]
Closing in favor of #136, which incorporates these changes in a slightly different way.

**Pull request merged**